File: 3º Semestre/Técnica de Programação II/Factory Pessoas/Prototype.py
import copy

class Pessoa:
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade
        
    def __str__(self):
        return f'nome: {self.nome}, idade: {self.idade}'
    
    def clone (self):
        return copy.copy(self)

class PessoaManager:
    
    def __init__(self):
        self.pessoas = {}
        
    def add_pessoa(self, nome, idade, id):
        pessoa = Pessoa(nome, idade)
        self.pessoas[id] = pessoa
        
    def get_pessoa(self, id):
        return self.pessoas[id].clone()
    
manager = PessoaManager()

manager.add_pessoa('Humpty', 10, 1)
manager.add_pessoa('Canalha', 20, 2)
manager.add_pessoa('Ovni', 30, 3)
manager.add_pessoa('Humpty dnv', 10, 4)
        
# Clonar

pessoa1 = manager.get_pessoa(1)
pessoa2 = manager.get_pessoa(2)
            
# Modificar informações

pessoa1.idade = 18
pessoa2.nome = "Gustavo Humpty"
pessoa2.idade = 80


# Imprimir resultados

print(manager.get_pessoa(1))
print(manager.get_pessoa(2))
print(manager.get_pessoa(3))
print(manager.get_pessoa(4))
print(pessoa1)
print(pessoa2)

File: 3º Semestre/Técnica de Programação II/Factory Pessoas/Prototype.js
class Pessoa{
    constructor(nome, idade){
        this.nome = nome;
        this.idade = idade;
    }

    clone(){
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }
}

class PessoaManager{
    constructor(){
        this.pessoas = {};
    }

    addPessoa(nome, idade, id){
        const pessoa = new Pessoa(nome, idade);
        this.pessoas[id] = pessoa;
    }

    getPessoa(id){
        return this.pessoas[id].clone();
    }   
    
}

const manager = new PessoaManager()

// Adicionando pessoas

manager.addPessoa('Pessoa 1', 10, 1);
manager.addPessoa('Pessoa 2', 20, 2);
manager.addPessoa('Pessoa 3', 30, 3);

// Clonando pessoas

const pessoa1 = manager.getPessoa(1);
const pessoa2 = manager.getPessoa(2);
const pessoa3 = manager.getPessoa(3);

// Modificar as pessoas clonadas

pessoa1.nome = 'Pessoa alterada 1'
pessoa2.nome = 'Pessoa alterada 2'
pessoa3.nome = 'Pessoa alterada 3'

// Imprimindo informações

console.log('Pessoas criadas\n');
console.log(manager.getPessoa(1));
console.log(manager.getPessoa(2));
console.log(manager.getPessoa(3));
console.log('\nPessoas clonadas\n');
console.log(pessoa1);
console.log(pessoa2);
console.log(pessoa3);